elastic/kibana
# [Fleet] Updates to output logic

Closes https://github.com/elastic/kibana/issues/152234

## Summary

- If the user changes an output from `ES` to `logstash` when there are Fleet server policies that use that output, we throw asking the user to create a new output 
- If the user changes the default output to a `logstash` one, we update the Fleet server policies to use the previous default output (that should be ES)
- Removed limitations to basic licenses so that fleet server policies can select a per-policy output (between the  available ES outputs). This means that "Basic" users should be able to set their default output to Logstash but still have the Fleet Server policy attached to an ES output.
- Added some tests to cover new functionalities

### UI changes

When editing a "logstash" output the modal displays a warning:

<img width="878" alt="Screenshot 2023-03-16 at 15 04 54" src="https://user-images.githubusercontent.com/16084106/225665515-6090b85c-b84f-4295-9b30-c480d2a84879.png">

### Testing

1. **Edit existing logstash output to be default**
- Have a default ES output and a fleet server policy (verify that has fleet server integration)
- Create a new logstash output 
- Update it to make it default (this is valid for both integrations and monitoring output)
- Check that fleet server policy keeps the ES output, it can be verified from agent policies > fleet server policy > settings. 
- The other policies should be switched to the logstash output

**Note** if the previous output is logstash, it will be switched to the default one. However, users who updated the system and already had a logstash ouput set as a default will need to to manually set the output from the agent policy settings page.

2. **Edit existing ES output to become a logstash one**
- Have a default ES output and a fleet server policy (verify that has fleet server integration)
- Edit the output and change it to type 'logstash' 
- Check that the endpoint throws with an error

<img width="366" alt="Screenshot 2023-03-21 at 18 06 25" src="https://user-images.githubusercontent.com/16084106/226687174-296e6e2b-3cef-4649-8173-18ee70c6a4a7.png">

3. **Create new logstash output as default**
- Have a default ES output and a fleet server policy (verify that has fleet server integration)
- Create a new logstash output as set it as default output
- Check that the fleet server policy keeps the  previous ES default output
- basically this should behave like case 1 (PUT and POST should have the same behaviour)

5. **Delete ES Output**
- Once in the previous case (fleet server on ES output, other policies on logstash), try to delete the ES output
- It should be prevented with an error

<img width="444" alt="Screenshot 2023-03-16 at 15 05 44" src="https://user-images.githubusercontent.com/16084106/226687206-f22e1330-4435-4077-a7ef-f29846f25e00.png">

6. **Custom output per policy**
- With "basic" license, go to an agent policy with fleet server > settings
- Check that "output for integrations" can select any output of type `elasticsearch`, while the logstash ones are disabled

<img width="981" alt="Screenshot 2023-03-21 at 18 09 29" src="https://user-images.githubusercontent.com/16084106/226687854-95d239a3-4122-470a-88e0-73c456ad5f7d.png">

### Checklist
